packages/paid-storage-contract/src/contract.ts

```diff
@@ -0,0 +1,192 @@
+    const amount: Balance = near.attachedDeposit();
+    account_id = account_id ?? near.predecessorAccountId();
+
+    assert(amount > 0, "Deposit amount must be greater than zero");
+
+    if (this.accounts.containsKey(account_id)) {
+      if (registration_only) {
+        near.log!("The account is already registered, refunding the deposit");
+        NearPromise.new(near.predecessorAccountId()).transfer(amount);
+      } else {
+        const currentValues = this.accounts.get(account_id);
+        this.accounts.set(
+          account_id,
+          new StorageBalance(
+            currentValues.total + amount,
+            currentValues.available + amount
+          )
+        );
+      }
+    } else {
+      let min_balance: Balance = this.storage_balance_bounds().min;
+      if (amount < min_balance) {
+        throw Error(
+          "The attached deposit is less than the minimum storage balance"
+        );
+      }
+
+      if (registration_only) {
+        this.accounts.set(account_id, new StorageBalance(0n, 0n));
+        let refund: Balance = amount - min_balance;
+
+        if (refund > 0) {
+          NearPromise.new(near.predecessorAccountId()).transfer(refund);
+        }
+      } else {
+        const initialAmount = amount - min_balance;
+        this.accounts.set(
+          account_id,
+          new StorageBalance(initialAmount, initialAmount)
+        );
+      }
+    }
+
+    return this.accounts.get(account_id);
+  }
+
+  @call({})
+  storage_withdraw({ amount }: { amount: Option<bigint> }): StorageBalance {
+    amount = BigInt(amount);
+    assert_one_yocto();
+
+    const predecessor_account_id = near.predecessorAccountId();
+
+    const storage_balance = this.internal_storage_balance_of(
+      predecessor_account_id
+    );
+
+    if (!storage_balance) {
+      throw Error(`The account ${predecessor_account_id} is not registered`);
+    }
+
+    if (amount > storage_balance.available) {
+      throw Error("The amount is greater than the available storage balance");
+    }
+
+    if (amount) {
+      this.accounts.set(
+        predecessor_account_id,
+        new StorageBalance(
+          storage_balance.total - amount,
+          storage_balance.available - amount
+        )
+      );
+    } else {
+      this.accounts.remove(predecessor_account_id);
+    }
+
+    NearPromise.new(predecessor_account_id).transfer(
+      amount || storage_balance.available
+    );
+
+    return this.accounts.get(predecessor_account_id);
+  }
+
+  storage_unregister({ force }: { force: boolean }): boolean {
+    assert_one_yocto();
+    const predecessor_account_id = near.predecessorAccountId();
+
+    const storage_balance = this.internal_storage_balance_of(
+      predecessor_account_id
+    );
+
+    if (!storage_balance) {
+      return false;
+    }
+
+    if (storage_balance.available && !force) {
+      throw Error(
+        `The account ${predecessor_account_id} has available balance!`
+      );
+    }
```

Comment:
This check was implemented according to the NEP note on `storage_unregister`:
```
// If `force=false` or `force` is omitted, the contract MUST panic if caller
// has existing account data, such as a positive registered balance (eg token
// holdings).
```

I think this implementation does not contradict with the idea:
```
logic should be that if a user has any tied up storage besides the registration min (meaning they have dapp code stored) then unregister fails unless force is provided, in which case force would also wipe their data
```
since `storage besides the registration min` is exactly what `.available` value represents.

**Update!**
After taking a second look at the NEP description of the method, I think that this line `// Unregisters the predecessor account and returns the storage NEAR deposit.` contradicts with the rest of the description, since f `force=true` it should burn balances, while `force=false` or `force` is omitted, the contract MUST panic. So it seems to me that there is no way when the storage NEAR deposit is being refunded.

Reply:
This confused me at first too. The positive registered balance is referring to fungible token balance, not storage balance. If force is true, then the users data is forcefully wiped even if they have things they might care about like fungible tokens, NFTS, or in our case dapp code

Reply:
I've updated the logic, so now if `force=false` or `force` is omitted, we do a refund of the account registration costs only. I've also added a check just in case the `total` is less than the current registration cost (but it might be a redundant one).
Technically it means that the user doing a force unregister will get a refund of the storage registration cost only and the available funds (if there are any) will stay on the contract balance.

```
const min_balance: Balance = this.storage_balance_bounds().min;
const refund = storage_balance.total >= min_balance ? min_balance : storage_balance.total;
NearPromise.new(predecessor_account_id).transfer(refund);
```

@mpeterdev is the anything we could improve here in your opinion?

Reply:
**UPDATE**

`storage_unregister` with `force=true` deletes the account record and refund the total user balance.